proposal: Uniform way for handling JWTs

kimwnasptd opened the pull request:
Continuing from the discussion in https://github.com/kubeflow/manifests/issues/2747

After the `oauth2-proxy` PRs are merged we have made some changes on how JWTs are used from within the cluster (`RequestAuthentication`) and how headers are added from these tokens.

This lead to some issues in KFP, for which to resolve we'll need to have an agreed upon story on how to be working in the future with JWTs.

After agreeing on the above, this can also enable us to push the effort for supporting groups since we'll have an agreed upon way of working with identity information.

/cc @juliusvonkohout @kromanow94 @thesuperzapper @axel7083 
cc @rimolive @DnPlas

DnPlas commented on proposals/20240606-jwt-handling.md:
```diff
@@ -0,0 +1,125 @@
+https://istio.io/latest/docs/tasks/security/authorization/authz-jwt/
+
+But the above creates the following issues:
+2. Ambiguity on how to handle different `id_tokens` (i.e. from Dex and from K8s)
```
I think this is not getting rendered correctly, [check the file](https://github.com/kubeflow/manifests/blob/0121ae865418aa6b81d3e6fa8740d765972bf5e8/proposals/20240606-jwt-handling.md#limitations)

kimwnasptd replied:
Ah good catch, thanks!

juliusvonkohout commented:
The general approach looks good, although i did not check the technical details yet.

kromanow94 commented:
Hey, this is the continuation of my thoughts from https://github.com/kubeflow/manifests/issues/2747#issuecomment-2168641503.

In general, I like this approach. In the comment linked above, I mentioned that I'd personally vote for not removing implementation of JWTs in Kubeflow components but after some thoughts over the weekend, I don't emphasize that so much right now.

While I personally would love Kubeflow to be natively supporting JWTs, I understand that we have to take into account the balance between functionality and maintenance/development and offload whatever possible. With this in mind, if you decide to drop JWT implementation from kubeflow components and just rely on the `kubeflow-userid` and `kubeflow-groups` headers and enforce those through Istio, I'm ok with that and you have my support.
  * If possible, I'd suggest standardizing the auth headers to `X-Auth-Request-Groups` and `X-Auth-Request-Email`.

I also mentioned that I'd prefer to specify `AuthorizationPolicy` objects with custom action and `oauth2-proxy` as the provider for every component that's exposed to users (api-server, jupyter-web-app, central-dashboard, etc...), not just for the istio-ingressgateway, to bring security closer to the components.
* While this would move the security layer closer to the components and allow increased flexibility in implementation of how the requests are routed and authorized to reach an endpoint, this is also not something I'm putting strong emphasis on. If you like that approach, I'd be happy to help with hands-on, but this is not something we have to have.
* So, to put it short: from in-cluster perspective, using a pair of `AuthorizationPolicy` and `RequestAuthentication` would be enough and the requests wouldn't have to go trough `ouath2-proxy`.

And then to continue some other thoughts:
* Since this proposal is suggesting removal of the JWT implementation from Kubeflow components, it also means that we don't need to check for `pipelines.kubeflow.org` audience and we don't have to add `PodDefaults` that adds the projected token.
  * I suggest to drop the `pipelines.kubeflow.org` audience at some form either way so I like that.

* I suggest dropping the `istio-ingressgateway-service-account` from the `AuthorizationPolicices` principal and just ensure that any principal is available through `rules.from.source.requestPrincipals`. This would also ease integration with externally installed Istio since the service account name for istio-ingressgateway depends on if you install through Helm Chart or the istio operator/istioctl.

kimwnasptd commented:
Thanks for the review @kromanow94! 

To avoid answering multiple parts in the same message, ending in big messages, could you provide also the above points as concrete comments on the proposal so we keep the discussions focused/separate?

kimwnasptd commented:
Thank you so much for the review comments @kromanow94! I'll take a look on them today/tomorrow

kimwnasptd commented on proposals/20240606-jwt-handling.md:
```diff
@@ -0,0 +1,125 @@
+## Scope
+This proposal aims to standardise how the Kubeflow backends should be handling the user information (name of user, groups they belong to), living in JWTs and http headers.
+
+This proposal takes as a requirement that users should be able to use K8s Tokens as `Authorization: Bearer <token>` headers in their request from inside the cluster.
```
> At end, K8s ServiceAccount Tokens, Dex Tokens, even GitLab Tokens are JWTs and the difference is only who provides those and what claims are available.

Yes! That's why I like the proposal of off-loading the JWT handling in the service-mesh. Because then the apps can be completely.

Pushed a commit to try and clarify this and the multiple issuers. LMK what you think
7a4aeb91

kromanow94 replied:
I like the updated version. Thanks.

kimwnasptd commented on proposals/20240606-jwt-handling.md:
```diff
@@ -0,0 +1,125 @@
+### In
+- Which component should be validating JWTs (id-tokens from OIDC or K8s ServiceAccount tokens)
+- Define where the backends should expect to find user related information
+- Define how different token "types" (dex, k8s) should be handled
```
Ah good point to clarify, I meant about different JWT **issuers**. At the end of the day what the token represents (user, k8s-service account, some vm) should be irrelevant to the apps.

Adding a small clarifications. I'll leave the comment open though in case you had further suggestions here
5a0b4e27

kromanow94 replied:
I like the updated version. Thanks.

kimwnasptd commented on proposals/20240606-jwt-handling.md:
```diff
@@ -0,0 +1,125 @@
+    2. The KFP backend [explicitly drops requests](https://github.com/kubeflow/manifests/blob/96ce068e16b2a707464471bddc0d2a58e403d1fc/apps/pipeline/upstream/base/installs/multi-user/istio-authorization-config.yaml#L37) from user namespaces if they have this header
+    3. For in-cluster Pods that want to talk to `kubeflow` workloads, which understands identity, the Pods will need to [use a K8s ServiceAccount Token](https://www.kubeflow.org/docs/components/pipelines/v1/sdk/connect-api/#full-kubeflow-subfrom-inside-clustersub)
+### Limitations
+- Not able to express `AuthorizationPolicies` for group header in Istio
```
> not able to use JWT Claims to configure access to kserve, limiting possibility to define m2m auth for model service

Regarding this one, can't users now (via AuthService) use K8s tokens for ISVCs?

Added a commit to include the second part though
5c694ba8

kromanow94 replied:
If they know how to handle that - sure. I was thinking we should support that scenario more, for example configure only KF Profile Contributors with access to a specific model. If this is not something we want to focus on then we can skip.

kimwnasptd commented on proposals/20240606-jwt-handling.md:
```diff
@@ -0,0 +1,125 @@
+### Limitations
+- Not able to express `AuthorizationPolicies` for group header in Istio
+
+The above limitation has pushed us to inject `id_tokens` to requests, via `oauth2-proxy`, to all authenticated requests. This will enable us to define policies in the future for better handling of groups.
```
Pushed a commit to rephrase, LMKWYT
0fd595bd

kromanow94 replied:
Nice, thanks.

kimwnasptd commented on proposals/20240606-jwt-handling.md:
```diff
@@ -0,0 +1,125 @@
+But the above creates the following issues:
+1. Ambiguity on how to handle different `id_tokens` (i.e. from Dex and from K8s)
+2. Information of user is both in `kubeflow-userid` and in `id_token` of http request
+3. It's not clear if backends should be validating the JWT tokens (i.e. KFP right now validates ServiceAccount tokens [`1`](https://github.com/kubeflow/pipelines/blob/2.2.0/backend/src/apiserver/auth/authenticator_token_review.go#L47-L58) [`2`](https://github.com/kubeflow/pipelines/blob/2.2.0/backend/src/apiserver/resource/resource_manager.go#L1698-L1699) )
```
Pushed a commit to only reference JWT, and not the extra token word. Feel free to resolve the converstation if that's all that you had in mind
fd947388

kimwnasptd commented on proposals/20240606-jwt-handling.md:
```diff
@@ -0,0 +1,125 @@
+2. The service-mesh is responsible for validating the JWTs
+    1. The service-mesh must drop (401) a request if the JWT is invalid (`RequestAuthentication`)
+    2. The service-mesh must drop (403) a request if the JWT is not present, and the application expects requests to have a user identity (`AuthorizationPolicy`)
+3. The backends are not responsible for validating the JWTs or their existence
```
The above indeed sound like interesting use cases for the future, but would also like us to evaluate this with a concrete proposal once the need will be more prominent for the above.

The argument for abstracting the service mesh especially is quite interesting, although would require us to iron out a lot of details.

Are you OK with keeping this comment open then, so that we reference it in the future discussions, but not block this PR on the above for now?

kromanow94 replied:
Sounds good man, let's do it like that.

kimwnasptd commented on proposals/20240606-jwt-handling.md:
```diff
@@ -0,0 +1,125 @@
+4. The service-mesh must expose user and groups to `kubeflow-userid` and `kubeflow-groups` headers, after validating JWTs
+    1. if the mesh will not override these headers, then users could forge requests and impersonate other users by setting the header and any valid token
+    1. by default the `email` claim from `id_token`, should be configurable in one place per issuer via `RequestAuthentication`
+    2. in case of K8s ServiceAccount tokens the `email` claim will be `system:serviceaccount:<sa-namespace>:<sa-name>`
```
Very good point!

kimwnasptd commented on proposals/20240606-jwt-handling.md:
```diff
@@ -0,0 +1,125 @@
+    2. in case of K8s ServiceAccount tokens the `email` claim will be `system:serviceaccount:<sa-namespace>:<sa-name>`
+    3. in case of Dex tokens it would be `kimonas@email.com`
+5. The backends will use the information from the headers and not deal with JWTs
+    1. `SubjectAccessReviews` should be made for the `user` and `groups` that were exposed from the headers, independently of the issuer of the token
```
Pushed a commit 
7ba68dec

kimwnasptd commented on proposals/20240606-jwt-handling.md:
```diff
@@ -0,0 +1,125 @@
+https://istio.io/latest/docs/tasks/security/authorization/authz-jwt/
+
+But the above creates the following issues:
+1. Ambiguity on how to handle different `id_tokens` (i.e. from Dex and from K8s)
```
@kromanow94 I agree to all the above, but I don't think all the above are immediately clear to a reader once they see "let's standardise JWTs" proposal.

That's why I mentioned it there, so that they are prepared that the proposal will go through this and explain that for Kubeflow these tokens are the same, independently of their issuer. And it's up to the service-mesh to handle the issuer.

Rephrased this to point out that these are topics that need to be cleared from the proposal
3849cf92

kimwnasptd commented on proposals/20240606-jwt-handling.md:
```diff
@@ -0,0 +1,125 @@
+
+But the above creates the following issues:
+1. Ambiguity on how to handle different `id_tokens` (i.e. from Dex and from K8s)
+2. Information of user is both in `kubeflow-userid` and in `id_token` of http request
```
Yes, rephrased the word limitation for https://github.com/kubeflow/manifests/pull/2748#discussion_r1647396547

kimwnasptd commented on proposals/20240606-jwt-handling.md:
```diff
@@ -0,0 +1,125 @@
+- By using `requestPrincipals`
+- By using `request.auth.claims[iss]` in the `when` condition of an `AuthorizationPolicy` rule
+
+The recommended way is to use `requestPrincipals: ["*"]`, as the [Istio docs suggest](https://istio.io/latest/docs/tasks/security/authorization/authz-jwt/), to accept only requests that have a JWT. But if a Kubeflow app would like to only accept tokens from a specific issuer (i.e. k8s) then the `request.auth.claims[iss]` key should be used.
```
Agree, pushed a commit [86df956](https://github.com/kubeflow/manifests/pull/2748/commits/86df95637315bce6fd1e1ad561242c7093740d4b)

kimwnasptd commented on proposals/20240606-jwt-handling.md:
```diff
@@ -0,0 +1,125 @@
+  - from:
+    - source:
+        requestPrincipals: # new! Allow request from any namespace, as long as it has JWT
+        - '*'
```
Pushed a commit a3c1ac4c

kimwnasptd commented:
@kromanow94 I've addressed the review comments, and only left the ones that propose the removal of the IngressGateway declaration in the AuthorizationPolicies.

I would like to avoid making this change in this PR. Mainly because this is a bigger discussion, to untangle the Istio IngressGateway, from the Kubeflow components. There are other places that are affected by this, like the AuthorizationPolicies created by Profiles (for allowing traffic for Notebooks and/or ISVCs).

So I'd prefer that we tackle that problem of its own, in a proposal that captures all the aspects rather than doing some changes here that are not fully spec-ed out for the overall effort.

WDYT?

juliusvonkohout commented:
Given oauth2-proxy the information in https://www.kubeflow.org/docs/components/pipelines/user-guides/core-functions/connect-api/#full-kubeflow-subfrom-inside-clustersub is is really misleading.

juliusvonkohout commented on proposals/20240606-jwt-handling.md:
```diff
@@ -0,0 +1,135 @@
+- Not able to express `AuthorizationPolicies` for group header in Istio
+- Limited possibility to use custom JWT claims as a source of information about authenticated user
+
+To accommodate the above limitations and improve the authnz flow in terms of security, maintenance and flexibility
```
To accommodate the above limitations and improve the authentication and authorization flow in terms of security,

kimwnasptd replied:
pushed a commit a4ba96a4

juliusvonkohout commented on proposals/20240606-jwt-handling.md:
```diff
@@ -0,0 +1,135 @@
+    3. In-cluster Pods that want to talk to `kubeflow` workloads, which understands identity, are [using a K8s ServiceAccount Token](https://www.kubeflow.org/docs/components/pipelines/v1/sdk/connect-api/#full-kubeflow-subfrom-inside-clustersub)
+### Limitations
+- Not able to express `AuthorizationPolicies` for group header in Istio
+- Limited possibility to use custom JWT claims as a source of information about authenticated user
```
about the authenticated user

kimwnasptd replied:
pushed a commit a4ba96a4

juliusvonkohout commented on proposals/20240606-jwt-handling.md:
```diff
@@ -0,0 +1,135 @@
+
+To accommodate the above limitations and improve the authnz flow in terms of security, maintenance and flexibility
+we propose to add the JWT to the `Authorization` header, so it can be digested by Istio and have the user details securely
+injected to auth headers when authorized. This will also enable us to define policies in the future for better handling of groups.
```
securely injected into the `Authorization` headers. Since you write "add the JWT to the `Authorization` header" in the previous sentence.

kimwnasptd replied:
pushed a commit a4ba96a4

juliusvonkohout commented on proposals/20240606-jwt-handling.md:
```diff
@@ -0,0 +1,135 @@
+injected to auth headers when authorized. This will also enable us to define policies in the future for better handling of groups.
+https://istio.io/latest/docs/tasks/security/authorization/authz-jwt/
+
+But the above creates the following topics that require agreement on how to handle:
```
require an agreement on how to handle them:

kimwnasptd replied:
pushed a commit a4ba96a4

juliusvonkohout commented on proposals/20240606-jwt-handling.md:
```diff
@@ -0,0 +1,135 @@
+3. It's not clear if backends should be validating the JWT (i.e. KFP right now validates ServiceAccount tokens [`1`](https://github.com/kubeflow/pipelines/blob/2.2.0/backend/src/apiserver/auth/authenticator_token_review.go#L47-L58) [`2`](https://github.com/kubeflow/pipelines/blob/2.2.0/backend/src/apiserver/resource/resource_manager.go#L1698-L1699) )
+
+## Spec
+This proposal aims to provide a uniform way for all backends to handle identity tokens, and agreeing on which levels of the stack should responsible for different parts.
```
## Specification
The goal of this proposal is to provide a uniform way for all backends to handle identity tokens and to specify which levels of the stack are responsible for which parts.

kimwnasptd replied:
pushed a commit a4ba96a4

juliusvonkohout commented on proposals/20240606-jwt-handling.md:
```diff
@@ -0,0 +1,135 @@
+    1. The service-mesh must drop (401) a request if the JWT is invalid (`RequestAuthentication`)
+    2. The service-mesh must drop (403) a request if the JWT is not present, and the application expects requests to have a user identity (`AuthorizationPolicy`)
+3. The backends are not responsible for validating the JWTs or their existence
+4. The service-mesh must expose user and groups to `kubeflow-userid` and `kubeflow-groups` headers, after validating JWTs
```
will it drop the bearer token header with the JWT before forwarding to the backend?

kimwnasptd replied:
No no the JWT will stay in the `Authorization` header, and the mesh will add the headers as an extra

juliusvonkohout commented on proposals/20240606-jwt-handling.md:
```diff
@@ -0,0 +1,125 @@
+### Implementation
+The technical details for the above proposal translate to the following
+1. Common Kubeflow manifests, for all components, for configuring Istio for supporting multiple issuers ([Dex](https://github.com/kubeflow/manifests/blob/v1.9-branch/common/oidc-client/oauth2-proxy/components/istio-external-auth/requestauthentication.dex-jwt.yaml) and [K8s-m2m](https://github.com/kubeflow/manifests/blob/v1.9-branch/common/oidc-client/oauth2-proxy/components/istio-m2m/requestauthe ...
+2. `AuthorizationPolicy` objects of components, for allowing access from Istio IngressGateway, will need to be extended for also requiring a JWT
```
" `AuthorizationPolicy` objects of components, for allowing access from Istio IngressGateway, must be adjusted to enforce the presence of a JWT"

kromanow94 replied:
Or maybe

"`AuthorizationPolicy` objects of KF Components have to be adjusted to allow access only when a trusted JWT is present in the `Authorization: Bearer <Token>` header. We no longer have to enforce access only through `istio-ingressgateway`."

juliusvonkohout commented on proposals/20240606-jwt-handling.md:
```diff
@@ -0,0 +1,135 @@
+6. `RequestAuthentication` objects, per issuer, should expose the corresponding token claims to the `kubeflow-userid` and `kubeflow-groups` headers
+7. Backends only need to care about `kubeflow-userid` and `kubeflow-groups` headers
+#### Requiring a JWT
+The service-mesh will need to drop requests (403) that don't have any JWT, for apps that expect user identity in the requests.
```
The service-mesh will need to drop requests (403) that don't have any JWT, for services that expect a user identity in the requests.

kimwnasptd replied:
pushed a commit a4ba96a4

juliusvonkohout commented on proposals/20240606-jwt-handling.md:
```diff
@@ -0,0 +1,135 @@
+#### Requiring a JWT
+The service-mesh will need to drop requests (403) that don't have any JWT, for apps that expect user identity in the requests.
+
+This can be achieved by multiple ways
```
This can be achieved in multiple ways:

kimwnasptd replied:
pushed a commit a4ba96a4

juliusvonkohout commented on proposals/20240606-jwt-handling.md:
```diff
@@ -0,0 +1,135 @@
+- By using `requestPrincipals`
+- By using `request.auth.claims[iss]` in the `when` condition of an `AuthorizationPolicy` rule
+
+The recommended way is to use `requestPrincipals: ["*"]`, as the [Istio docs suggest](https://istio.io/latest/docs/tasks/security/authorization/authz-jwt/), to accept only requests that have a JWT.
```
that have a ?valid? JWT.

kimwnasptd replied:
pushed a commit a4ba96a4

juliusvonkohout commented on proposals/20240606-jwt-handling.md:
```diff
@@ -0,0 +1,135 @@
+
+The recommended way is to use `requestPrincipals: ["*"]`, as the [Istio docs suggest](https://istio.io/latest/docs/tasks/security/authorization/authz-jwt/), to accept only requests that have a JWT.
+
+If an admin would like to further limit access to Kubeflow apps based on specific issuers, they can do so by updating the `AuthorizationPolicies`
```
Kubeflow services based

kimwnasptd replied:
pushed a commit a4ba96a4

kromanow94 commented:
@kimwnasptd 
> @kromanow94 I've addressed the review comments, and only left the ones that propose the removal of the IngressGateway declaration in the AuthorizationPolicies.
> 
> I would like to avoid making this change in this PR. Mainly because this is a bigger discussion, to untangle the Istio IngressGateway, from the Kubeflow components. There are other places that are affected by this, like the AuthorizationPolicies created by Profiles (for allowing traffic for Notebooks and/or ISVCs).
> 
> So I'd prefer that we tackle that problem of its own, in a proposal that captures all the aspects rather than doing some changes here that are not fully spec-ed out for the overall effort.
> 
> WDYT?

Ok, this makes very much sense. Let's do it like that.

Also, I wasn't able to resolve the threads but I either left a comment or a thumbs up where I agree on the resolution.

---
@juliusvonkohout 
> Given oauth2-proxy the information in https://www.kubeflow.org/docs/components/pipelines/user-guides/core-functions/connect-api/#full-kubeflow-subfrom-inside-clustersub is is really misleading.

So I understand you suggest to update this docs with information on how to use the SA Token?

---

BTW, since we're moving around some stuff for authz, maybe you're interested in having a service that will manage programmatic login? This would be something that replaces the [Full Kubeflow (from outside cluster)](https://www.kubeflow.org/docs/components/pipelines/user-guides/core-functions/connect-api/#full-kubeflow-subfrom-outside-clustersub) flow. We have this little component that works with a websocket, can redirect to the auth page, get the cookies or a token from login response and send back via websocket to the client. We can also have a call sometime next week so I can showcase how would that work.

juliusvonkohout commented:
> So I understand you suggest to update this docs with information on how to use the SA Token?

Yes :-) this session hack i snot what we should advertise.

kimwnasptd commented:
@kromanow94 @juliusvonkohout I've done a pass and integrated your comments. Let me know if there's anything else

juliusvonkohout commented:
> @kromanow94 @juliusvonkohout I've done a pass and integrated your comments. Let me know if there's anything else

For me it is anyway a

/lgtm

for a long time (despite the "nits" :-D)

We can always reiterate and reiterating is easier if it is merged first.

juliusvonkohout commented:
/approve
